drivers: digital-io: max22190 : Add max22190 driver support

RaduSabau1 opened the pull request:
## Pull Request Description

MAX22190 is an octal industrial digital input device with diagnostics that communicates through a SPI interface.
The files added by this PR have been tested with the MAX22190 PMB connected to a max32690 microcontroller.
The driver has the basic api for init, remove, register read, write and update, but also a custom CRC5 algorithm specific for this device 
alongside its specific api(filter_get, filter_set, chan_state).

## PR Type
- [ ] Bug fix (change that fixes an issue)
- [x] New feature (change that adds new functionality)
- [ ] Breaking change (has dependencies in other repos or will cause CI to fail)

## PR Checklist
- [x] I have followed the [Coding style guidelines](http://analogdevicesinc.github.io/no-OS/drivers_guide.html#coding-style)
- [ ] I have performed a self-review of the changes
- [x] I have commented my code, at least hard-to-understand parts
- [ ] I have build all projects affected by the changes in this PR
- [ ] I have tested in hardware affected projects, at the relevant boards
- [x] I have signed off all commits from this PR
- [ ] I have updated the documentation (wiki pages, ReadMe etc), if applies

buha commented on drivers/digital-io/max22190/iio_max22190.h:
```diff
@@ -0,0 +1,81 @@
+#include "max22190.h"
+
+struct max22190_ch_filter {
+	uint32_t wbe;
```
please document in a comment what this acronyms stand for: `wbe` and `fbp`

RaduSabau1 replied:
Will document them

RaduSabau1 replied:
Actually, struct max22190_ch_filter will be removed, because of it being used only in the specific application example i created to test this on.
In case of a default mode filters are bypassed, therefore this struct wouldn't make sense.

buha commented on drivers/digital-io/max22190/iio_max22190.c:
```diff
@@ -0,0 +1,608 @@
+		.shared = IIO_SHARED_BY_ALL
+	},
+	{
+		.name = "fault1",
```
as discussed, this approach is a bit ugly but iio doesn't provide a proper way for exposing the status that comes with the data, so from my point of view it's good, maybe team has a better idea ?

RaduSabau1 replied:
Will try this instead.

buha commented on drivers/digital-io/max22190/iio_max22190.c:
```diff
@@ -0,0 +1,608 @@
+ * 	  channel
+ *
+ * @param dev - The iio device structure
+ *
```
idk about these blank lines... i think we'd be better without

RaduSabau1 replied:
Will remove them.

CiprianRegus commented on drivers/digital-io/max22190/iio_max22190.h:
```diff
@@ -0,0 +1,81 @@
+struct max22190_iio_desc {
+	struct max22190_desc *max22190_desc;
+	struct iio_device *iio_dev;
+	uint32_t active_channels;
```
Unused

RaduSabau1 replied:
Will remove

CiprianRegus commented on drivers/digital-io/max22190/iio_max22190.h:
```diff
@@ -0,0 +1,81 @@
+	struct iio_device *iio_dev;
+	uint32_t active_channels;
+	uint32_t no_of_active_channels;
+	struct max22190_ch_filter ch_filter[MAX22190_CHANNELS];
```
Why do you need to store this?

RaduSabau1 replied:
I was using it in an application example, but by taking a closer look, i realize this should be project specific instead of it being implemented in the driver itself.

CiprianRegus commented on drivers/digital-io/max22190/iio_max22190.h:
```diff
@@ -0,0 +1,81 @@
+*/
+struct max22190_iio_desc_init_param {
+	struct max22190_init_param *max22190_init_param;
+	struct max22190_ch_filter ch_filter[MAX22190_CHANNELS];
```
You're not using this.

CiprianRegus commented on drivers/digital-io/max22190/iio_max22190.c:
```diff
@@ -0,0 +1,608 @@
+	uint32_t val;
+	int ret;
+
+	ret = max22190_reg_read(desc->max22190_desc, MAX22190_DIGITAL_INPUT_REG,
```
This is not for a specific channel, you're just storing the value of a register.

RaduSabau1 replied:
Will use no_os_field_get, my bad.

CiprianRegus commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,341 @@
+int max22190_filter_set(struct max22190_desc *desc, uint32_t ch, uint32_t wbe,
+			uint32_t fbp, enum max22190_delay delay)
+{
+	if (ch > MAX22190_CHANNELS)
```
off by 1. Check throughout the file.

RaduSabau1 replied:
will fix it.

CiprianRegus commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,341 @@
+
+	if (desc->crc_en) {
+		crc = max22190_crc(&desc->buff[0]);
+		if (crc != (desc->buff[2] & 0x1F))
```
What is `0x1F`

CiprianRegus commented on drivers/digital-io/max22190/max22190.h:
```diff
@@ -0,0 +1,129 @@
+	struct no_os_spi_desc *comm_desc;
+	struct no_os_gpio_desc *en_gpio;
+	uint8_t buff[MAX22190_FRAME_SIZE + 1];
+	enum max22190_ch_state channels[MAX22190_CHANNELS];
```
You're not using it.

RaduSabau1 replied:
I am using it in the max22190_chan_state function.

CiprianRegus replied:
It's better to not duplicate your device's state in the driver. Just read the status register each time.

CiprianRegus commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,341 @@
+			goto gpio_err;
+	}
+
+	ret = max22190_reg_update(descriptor, MAX22190_FAULT1_REG, NO_OS_BIT(6),
```
What does this do? Adding a macro for the mask may help.

RaduSabau1 replied:
A reg_update isn't really needed on that specific mask, so I will replace it with a simple register read to clear the latched faults.

RaduSabau1 commented:
V2

- Removed unused variables.
- Modified driver init function.
- fault1 and fault2 are debug attributes now.
- Fixed read_raw value to be masked for the specific channel.
- Addressed other comments.

RaduSabau1 commented:
V3

- POR bit is now cleared on initialization(fault1 register is read because of the reg_update api call in the init).

buha commented on drivers/digital-io/max22190/max22190.h:
```diff
@@ -0,0 +1,130 @@
+#define MAX22190_WIRE_BREAK_REG		0x0
+#define MAX22190_DIGITAL_INPUT_REG	0x2
+#define MAX22190_FAULT1_REG		0x4
+#define MAX22190_FILTER_IN_REG(x)	(0x6 + (2*x))
```
paranthesize x as `(x)`

buha replied:
missing whitespaces

RaduSabau1 commented:
V4

- Fixed MACRO definition bug.

buha commented on drivers/digital-io/max22190/README.rst:
```diff
@@ -0,0 +1,149 @@
+	struct no_os_spi_init_param max22190_spi_ip = {
+		.device_id = SPI_DEVICE_ID,
+		.extra = SPI_EXTRA,
+		.max_speed_hz = SPI_BAUDRATE,
```
please put hardcoded values here, proper examples, like you did for 14916

RaduSabau1 commented:
V5

- Updated README.

buha commented on drivers/digital-io/max22190/README.rst:
```diff
@@ -0,0 +1,149 @@
+-----------------
+
+MAX22190 has a total of 6 attributes:
+* ``raw - the state of a channel``
```
This doesn't render correctly, please test sphinx output

Also, you don't need to backtick quote. Perhaps use it on the attribute name to render like this (but in sphinx of course):
* `raw` - raw value of a channel

RaduSabau1 replied:
will test the sphinx output, and make sure it renders correctly.

RaduSabau1 commented:
V6

- Updated README.

RaduSabau1 commented:
V7

- Updated IIO driver(switched to channel's address instead of ch_num).

rbolboac commented on drivers/digital-io/max22190/README.rst:
```diff
@@ -0,0 +1,148 @@
+MAX22190 no-OS driver
```
you should also source this file in sphinx/source similarly to what it is already done for other drivers

RaduSabau1 commented:
V8

- Included README.rst in sphinx/source.

amiclaus commented on doc/sphinx/source/drivers_doc.rst:
```diff
@@ -18,6 +18,13 @@ ANALOG TO DIGITAL CONVERTERS
+DIGITAL INPUT/OUTPUT
+==========================
+.. toctree::
+   :maxdepth: 1
```
after the merge of #1955, you need only `drivers/max22190` to add, since the ToC is already created. (therefore the newly appeared conflict.

RaduSabau1 replied:
I thought about this when writing the code. Thinking of waiting for #1955 to be merged first, and then only add drivers/max22190 line in this file (since ToC will already be created).

amiclaus replied:
I merged it so you can resolve already the conflict :)

RaduSabau1 replied:
On it.

rbolboac commented on drivers/digital-io/max22190/iio_max22190.h:
```diff
@@ -0,0 +1,71 @@
+/** Free resources allocated by the initalization function. */
+int max22190_iio_remove(struct max22190_iio_desc *);
+
+#endif /* IIO_MAX22190_H */
```
missing new line at the end of file, other than that can be merged

RaduSabau1 commented:
V10

- Added new line at the EOF (iio_max22190.h).

CiprianRegus commented on drivers/digital-io/max22190/README.rst:
```diff
@@ -0,0 +1,148 @@
+Device Attributes
+-----------------
+
+MAX22190 has a total of 6 attributes:
```
These are channel attributes

CiprianRegus commented on drivers/digital-io/max22190/iio_max22190.c:
```diff
@@ -0,0 +1,544 @@
+	},
+	{
+		.name = "fault2",
+		.show = max22190_iio_read_fault2,
```
Use just 1 function and set different `priv` values for each of these struct elements.

RaduSabau1 replied:
Will do.

CiprianRegus commented on drivers/digital-io/max22190/iio_max22190.c:
```diff
@@ -0,0 +1,544 @@
+	if (!max22190_iio_channels)
+		return -ENOMEM;
+
+	for (i = 0; i < MAX22190_CHANNELS; i++) {
+		if (ch_enabled[i]) {
+			max22190_iio_channels[ch_offset] = (struct iio_channel)MAX22190_CHANNEL(i);
+			max22190_iio_channels[ch_offset].attributes = max22190_attrs;
+			max22190_iio_channels[ch_offset].ch_out = 0;
+			ch_offset++;
+		}
+	}
```
Is there any reason why wouldn't all channels be enabled?

RaduSabau1 replied:
No, will remove the boolean array and everything related to it.

RaduSabau1 commented:
V12

- Fixed README.

RaduSabau1 commented:
V13

- iio_channels are now declared static and iio_setup_channels is removed.

RaduSabau1 commented:
V14

- Fixed error for iio_driver in the iio_device structure.
- Fixed warnings

rbolboac commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,517 @@
+	int ret;
+	uint32_t reg_val;
+
+	switch(fault_enable) {
```
I think this switch case can be rewritten with less lines in a more generic way.
something like:
```
case MAX22190_FAULT1_WBGE ... MAX22190_FAULT1_FAULT2E:
ret = max22190_reg_read(desc, MAX22190_FAULT1_EN_REG, &reg_val);
		if (ret)
			return ret;
		*enabled = no_os_field_get(MAX22190_FAULT_MASK(fault_enable), reg_val);
		break;
```
where
```
#define MAX22190_FAULT_MASK(x) NO_OS_BIT(x)
```

similarly for fault 2, you will only have to take care of the enum offset:
```
case MAX22190_FAULT2_RFWBSE ... MAX22190_FAULT2_FAULT8CKE:
                ret = max22190_reg_read(desc, MAX22190_FAULT2_EN_REG, &reg_val);
		if (ret)
			return ret;
		*enabled = no_os_field_get(MAX22190_FAULT_MASK(fault_enable-MAX22190_FAULT2_RFWBSE), reg_val);
		break;
```

rbolboac commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,517 @@
+ * @param enabled - State of the selected enable (0 for disabled, 1 for enabled)
+ * @return 0 in case of succes, negative error code otherwise
+*/
+int max22190_fault_enable_get(struct max22190_desc *desc,
```
let's add similar APIs for other registers (e.g. config register, filters, etc)

RaduSabau1 replied:
Config reg has 2 bits that can be added to fault and 1 for register, and I will do exactly that.

RaduSabau1 commented:
V15

- Added API for fault_enable and wbe.
- Added debug attributes for fault_enable in the IIO driver.
- Modified README.rst.

RaduSabau1 commented:
V16

- Modified documentation to match the driver.

rbolboac commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,502 @@
+	return max22190_reg_update(desc, MAX22190_FILTER_IN_REG(ch),
+				   MAX22190_FAULT2_WBE_MASK,
+				   no_os_field_prep(MAX22190_FAULT2_WBE_MASK,
+						   enabled));;
```
remove extra semicolon

rbolboac commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,502 @@
+	int ret;
+
+	switch(fault_enable) {
+	case MAX22190_FAULT1_WBGE:
+	case MAX22190_FAULT1_24VME:
+	case MAX22190_FAULT1_24VLE:
+	case MAX22190_FAULT1_ALRMT1E:
+	case MAX22190_FAULT1_ALRMT2E:
+	case MAX22190_FAULT1_FAULT2E:
```
can use `...` syntax similarly to get function

rbolboac commented on drivers/digital-io/max22190/iio_max22190.c:
```diff
@@ -0,0 +1,622 @@
+				    const struct iio_ch_info *channel,
+				    intptr_t priv)
+{
+	int32_t val = 0;
```
if the offset is always 0, why have it at all?

rbolboac commented on drivers/digital-io/max22190/iio_max22190.c:
```diff
@@ -0,0 +1,622 @@
+		if (val == max22190_delay_avail[i])
+			break;
+
+		if (i == NO_OS_ARRAY_SIZE(max22190_delay_avail) - 1)
+			return -EINVAL;
```
code can be optimized by taking this if condition outside of for loop

RaduSabau1 commented:
V17

- Optimized code.
- Removed offset.
- Addressed other comments.

buha commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,492 @@
+					   MAX22190_FAULT_MASK(fault_enable),
+					   no_os_field_prep(MAX22190_FAULT_MASK(fault_enable),
+							   enabled));
+	case MAX22190_FAULT2_RFWBSE ... MAX22190_FAULT2_FAULT8CKE:
```
so what if I want to enabled MAX22190_FAULT2_RFWBSE ?
`MAX22190_FAULT1_FAULT2E` should probably be automatically enabled by the driver, it is currently not

buha replied:
This would imply keeping track of fault2 enables somewhere in desc->fault2en or something like that.
When any fault2 enable gets enabled, `MAX22190_FAULT1_FAULT2E` should be set.
When no fault2 enable remains, `MAX22190_FAULT1_FAULT2E` should be cleared.

RaduSabau1 replied:
I'll do that but having a fault2en in the desc isn't necessary.

RaduSabau1 commented:
V18

- Checking fault2 enable in fault1 register when setting fault 2 enable.

RaduSabau1 commented:
V19

- Added fault2en in device descriptor and modified fault_enable set API in order to optimize it.

buha commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,516 @@
+
+		desc->fault2en[fault_enable - MAX22190_FAULT2_RFWBSE] = enabled;
+
+		for (i = 0; i < MAX22190_FAULT2_ENABLES; i++)
```
probably works as-is but if fault2en would be a bitmask, you would not need this for loop

RaduSabau1 commented:
V20

- Modified fault2en to be a bit mask.

RaduSabau1 commented on drivers/digital-io/max22190/max22190.c:
```diff
@@ -0,0 +1,520 @@
+
+	no_os_spi_remove(desc->comm_desc);
+
+	if (desc->en_gpio)
```
it's not, it will be removed.

RaduSabau1 commented:
V21

- Removed unnecessary lines.

RaduSabau1 commented:
V21

- Removed unnecessary lines.
